### ISSUE TITLE
Incorrect SQL generated

### ISSUE THREAD

**Issue by uehlbran** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1669]
Pomelo is not correctly generating SQL code. I am using EF Core version 6.0.6 as well as Pomelo 6.0.1. The following code produces the following SQL:

`      SELECT `t`.`UserId`, `t`.`Username`, `o`.`OrderId`, `o`.`Total`, `o`.`UserId`
      FROM (
          SELECT `u`.`UserId`, `u`.`Username`
          FROM `Users` AS `u`
          WHERE `u`.`UserId` = 1
          LIMIT 1
      ) AS `t`
      LEFT JOIN `Orders` AS `o` ON `t`.`UserId` = `o`.`UserId`
      ORDER BY `t`.`UserId`
`

There is absolutely no reason that a second select statement is being generated. It is redundant and as far as I am aware it is a performance problem. It should be generating the following SQL code:

`SELECT u.userid, u.username, o.orderid, o.total, o.userid from users as u join orders as o on o.userid = u.userid`

C# Dummy code:

`using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Query;
using Microsoft.Extensions.Logging;
using Microsoft.VisualBasic.CompilerServices;

namespace LearningCSharp1
{
    class Program
    {
        public static async Task Main(string[] args)
        {
            var context = new Context();

            var user = await context.Users
                .Where(u => u.UserId == 1)
                .Include(u => u.Order)
                .FirstAsync();
        }
    }

    class User
    {
        public int UserId { get; set; }
        public string Username { get; set; } = null!;
        public IEnumerable<Order> Order { get; set; } = new List<Order>();

        public override string ToString()
        {
            return $"Id: {UserId} | Username: {Username}";
        }
    }

    class Order
    {
        public int OrderId { get; set; }
        public decimal Total { get; set; }
        public User User { get; set; } = null!;
        public int UserId { get; set; }
    }

    class Context : DbContext
    {
        public DbSet<User> Users => Set<User>();
        public DbSet<Order> Orders => Set<Order>();

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                .UseMySql(
                    "Server=localhost;Database=efcore;User=root;Password=password.",
                    ServerVersion.AutoDetect("Server=localhost;Database=efcore;User=root;Password=password.")
                )
                .UseQueryTrackingBehavior(QueryTrackingBehavior.NoTracking)
                .LogTo(Console.WriteLine, LogLevel.Information)
                .EnableSensitiveDataLogging();
        }
    }
}`

**Comment by uehlbran** [comment]
I tried moving the include but, it still generates the same SQL.

> @uehlbran Move the `Include()` call in front of the `Where()` call and check whether it generates the correct query now:
> 
> ```cs
> var user = await context.Users
>                 .Include(u => u.Order)
>                 .Where(u => u.UserId == 1)
>                 .FirstAsync();
> ```

**Comment by uehlbran** [comment]
@lauxjpn I finally got around to trying another provider and different database and I get the same SQL output so I don't think this is actually an issue with Pomelo but, rather an issue with entity framework. I guess EF core 6 still generates what I'd consider terrible SQL.

**Comment by smitpatel** [comment]
> Why do both single and first produce the same SQL?

They don't produce same SQL either. The LIMIT in both queries is different number. 

Now I realized why perf numbers are that way @lauxjpn . Both are running different SQL which generates different results.
MultiSelect version takes first User and associated orders which can be one or more so result set could have 5 rows too. SingleSelect is doing Limit 1 on top level so it will only take first User and first matching order and discard other orders if they are matching and should be included in the result when populating Orders collection via EF. Hence it is only selecting 1 row and not 5. Hence results makes sense, why SingleSelect is faster.

**Comment by uehlbran** [comment]
@smitpatel Ok. I misspoke. It's not the exact same SQL. I should have said both queries produce bad SQL code. I suppose I can understand why first might use a second select statement but, there is absolute no reason that single should. It's completely unnecessary. Every additional include would result in an additional select statement which makes no sense.

**Comment by lauxjpn** [comment]
> Now I realized why perf numbers are that way @lauxjpn . Both are running different SQL which generates different results.
> MultiSelect version takes first User and associated orders which can be one or more so result set could have 5 rows too. SingleSelect is doing Limit 1 on top level so it will only take first User and first matching order and discard other orders if they are matching and should be included in the result when populating Orders collection via EF. Hence it is only selecting 1 row and not 5. Hence results makes sense, why SingleSelect is faster.

@smitpatel Good catch, that's it! And that is why the statement needs to be generated by EF Core the way it already currently is, since otherwise only the first order entity would be returned, which would be an incorrect translation of the LINQ query.

So looks good and works as expected.
